### PR TITLE
refactor: introduce data store and refine s3 reader

### PR DESCRIPTION
**Pull request by xx01cyx**
1. Introduce `DataStore` for different underlying data caches. See comments for more.
2. Refactor the storage manager. Use the interface of `DataStore` to make the code cleaner.
3. Remove `chunk_size` in `S3ReaderStream`. It now returns the data it polls from S3 directly instead of storing the data in an internal buffer and returning the data by chunk size.
4. Separate `DiskReadStream` and other types of storage streams (e.g. `S3ReaderStream`, `MockS3ReaderStream`). This is due to the inherent difference between disk read and S3 read, where the former prefers reading into a fixed in-memory buffer and the latter prefers separate in-memory buffers.
5. Introduce `RandomDiskStream` for `MockS3Reader` to mock S3's behavior, where a **random** number of bytes is returned from the network.

### CODE REVIEW

File: storage-node/src/cache/data/disk.rs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+            }
+        });
+        Ok(Some(rx))
+    }
+
+    async fn write_data(&self, key: String, data: StorageReaderStream) -> ParpulseResult<usize> {
+        // NOTE(Yuanxin): Shall we spawn a task to write the data to disk?
+        let bytes_written = self
+            .disk_manager
+            .write_stream_reader_to_disk(data, &key)
```
Comment (lanlou1554) [inline storage-node/src/cache/data/disk.rs:63]: TODO: In the future, do we need to write to disk & send to client at the same time?

File: storage-node/src/disk/stream.rs [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+whatever manliness or humour my life.
+I offer you the loyalty of a man who has never
+been loyal.
+I offer you that kernel of myself that I have saved,
+somehow-the central heart that deals not
+in words, traffics not with dreams, and is
+untouched by time, by joy, by adversities.
+I offer you the memory of a yellow rose seen at
+sunset, years before you were born.
+I offer you explanations of yourself, theories about
```
Comment (lanlou1554) [inline storage-node/src/disk/stream.rs:214]: WOWWWW

**Review by lanlou1554 — APPROVED** [review]
LGTM

File: storage-node/src/cache/data/disk.rs (hunk 1 above)
Comment (xx01cyx) [reply storage-node/src/cache/data/disk.rs:63]: I think yes

**Comment by lanlou1554** [comment]
Wait, I find in MemDiskStore, memory and disk should have 2 different lru, which is different from DiskStore. Can I further refactor to put lru from storage_manager to diskstore?

**Pull request merged**